domains/zeto/internal/zeto/nonfungible/handler_transfer.go

```diff
@@ -0,0 +1,380 @@
+	}
+
+	if len(req.InputStates) != 1 {
+	}
+
+	// set input
+	input, err := prepareState(ctx, req.InputStates[0])
+	if err != nil {
+		return nil, err
+	}
+
+	// set output
+	output, err := prepareState(ctx, req.OutputStates[0])
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := encodeTransactionDataFunc(ctx, req.Transaction, types.ZetoTransactionData_V0)
+	if err != nil {
+		return nil, i18n.NewError(ctx, msgs.MsgErrorEncodeTxData, err)
+	}
+	params := map[string]any{
+		"input":  input,
+		"output": output,
+		"proof":  encodeProofFunc(proofRes.Proof),
+		"data":   data,
+	}
+	transferFunction := getTransferABI(tx.DomainConfig.TokenName)
+
+	if common.IsNullifiersToken(tx.DomainConfig.TokenName) {
+		delete(params, "input")
+		params["nullifier"] = strings.Split(proofRes.PublicInputs["nullifier"], ",")
+		params["root"] = proofRes.PublicInputs["root"]
+	}
+	paramsJSON, err := json.Marshal(params)
+	if err != nil {
+		return nil, i18n.NewError(ctx, msgs.MsgErrorMarshalPrepedParams, err)
+	}
+	functionJSON, err := json.Marshal(transferFunction)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.PrepareTransactionResponse{
+		Transaction: &pb.PreparedTransaction{
+			FunctionAbiJson: string(functionJSON),
+			ParamsJson:      string(paramsJSON),
+		},
+	}, nil
+}
+
+func prepareState(ctx context.Context, state *pb.EndorsableState) (string, error) {
+	token, err := makeNFToken(state.StateDataJson)
+	if err != nil {
+		return "", i18n.NewError(ctx, msgs.MsgErrorParseInputStates, err)
+	}
+	hash, err := token.Hash(ctx)
+	if err != nil {
+		return "", i18n.NewError(ctx, msgs.MsgErrorHashInputState, err)
+	}
+	return hash.String(), nil
+
+}
+
+func (h *transferHandler) formatProvingRequest(ctx context.Context, input, output []*types.ZetoNFToken, circuitId, tokenName, stateQueryContext string, contractAddress *tktypes.EthAddress) ([]byte, error) {
+
+	inputCommitments, inputSalts, tokenURIs, tokenIDs, inputOwners, err := processTokens(ctx, input)
+	if err != nil {
+		return nil, err
+	}
+
+	outputCommitments, outputSalts, _, _, outputOwners, err := processTokens(ctx, output)
+	if err != nil {
+		return nil, err
+	}
+
+	/* TODO: Add support for nullifiers
+	if common.IsNullifiersCircuit(circuitId) {
+		proofs, extrasObj, err := generateMerkleProofs(ctx, h.callbacks, h.merkleTreeRootSchema, h.merkleTreeNodeSchema, tokenName, stateQueryContext, contractAddress, inputCoins)
+		if err != nil {
+			return nil, i18n.NewError(ctx, msgs.MsgErrorGenerateMTP, err)
+		}
+		for i := len(proofs); i < inputSize; i++ {
+			extrasObj.MerkleProofs = append(extrasObj.MerkleProofs, &smt.Empty_Proof)
+			extrasObj.Enabled = append(extrasObj.Enabled, false)
+		}
+		protoExtras, err := proto.Marshal(extrasObj)
+		if err != nil {
+			return nil, i18n.NewError(ctx, msgs.MsgErrorMarshalExtraObj, err)
+		}
+		extras = protoExtras
+	}
+	*/
+
+	provingRequestExtras := &corepb.ProvingRequestExtras_NonFungible{
+		TokenIds:  tokenIDs,
```

Comment:
to be consistent, should move the `InputValues` and `OutputValues` to a new protobuf `ProvingRequestExtras_Fungible`. or I think a better approach is doing the following:
- add a new property in the `ProvingRequestCommon` definition called "tokenType", it can be "fungible" or "non-fungible"
- add a property of type `bytes` and call it "tokenSecrets". it can be unmarshal'ed into either of the following protobuf definitions, depending on the value of the token type above:
  - `TokenSecrets_Fungible`, which has `InputValues` and `OutputValues`
  - `TokenSecrets_NonFungible`, which has `TokenIds` and `TokenUris`